src/jit/codegenxarch.cpp

```diff
@@ -1152,6 +1152,39 @@ void CodeGen::genReturn(GenTreePtr treeNode)
     // Also, there is not much to be gained by materializing it as an explicit node.
     if (compiler->compCurBB == compiler->genReturnBB)
     {
+#if defined(_TARGET_AMD64_)
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+        if (compiler->compMethodReturnsMultiRegRetType())
+        {
+            CORINFO_CLASS_HANDLE retClsHnd = compiler->info.compMethodInfo->args.retTypeClass;
+            ReturnTypeDesc retTypeDesc;
+            retTypeDesc.InitializeStructReturnType(compiler, retClsHnd);
+
+            unsigned regCount = retTypeDesc.GetReturnRegCount();
+
+            if (varTypeIsGC(compiler->info.compRetType))
+            {
+                for (unsigned i = 0; i < regCount; ++i)
+                {
+                    gcInfo.gcMarkRegPtrVal(retTypeDesc.GetABIReturnReg(i), retTypeDesc.GetReturnRegType(i));
+                }
+            }
+
+            genProfilingLeaveCallback();
+
+
+            if (varTypeIsGC(compiler->info.compRetType))
+            {
+                for (unsigned i = 0; i < regCount; ++i)
+                {
+                    gcInfo.gcMarkRegSetNpt(retTypeDesc.GetABIReturnReg(i));
+                }
+            }
+            return;
```

Comment:
>           return; [](start = 1, length = 18)

returns in the middle of the method are very hard to notice and could lead to accidental bugs.  Instead I would suggest this to restructure this logic as follows.  Note that the below logic requires no #ifdefs.

e.g.
        ReturnTypeDesc retTypeDesc;
        if (compiler->compMethodReturnsMultiRegRetType())
        {
            CORINFO_CLASS_HANDLE retClsHnd = compiler->info.compMethodInfo->args.retTypeClass;
            retTypeDesc.InitializeStructReturnType(compiler, retClsHnd);
        }

        if (varTypeIsGC(compiler->info.compRetType))
        {
            gcInfo.gcMarkRegPtrVal(REG_INTRET, compiler->info.compRetType);
        }
        else if (compiler->compMethodReturnsMultiRegRetType())
        {
              for (unsigned i = 0; i < retTypeDesc.GetReturnRegCount(); ++i)
                {
                    gcInfo.gcMarkRegPtrVal(retTypeDesc.GetABIReturnReg(i), retTypeDesc.GetReturnRegType(i));
                }
        }

       genProfilingLeaveCallback();

        if (varTypeIsGC(compiler->info.compRetType))
        {
            gcInfo.gcMarkRegSetNpt(REG_INTRET);
        }
       else  if (compiler->compMethodReturnsMultiRegRetType())
       {
                for (unsigned i = 0; i < retTypeDesc.GetReturnRegCount(); ++i)
                {
                    gcInfo.gcMarkRegSetNpt(retTypeDesc.GetABIReturnReg(i));
                }
       }